File: 最大单词长度乘积.py
class Solution:
    def maxProduct(self, words: List[str]) -> int:
        n, ans = len(words), 0
        mask = [0 for _ in range(n)]
        for i in range(n):
            for c in words[i]:
                mask[i] |= (1 << (ord(c) - ord('a')))
        for i in range(n):
            for j in range(n):
                if mask[i] & mask[j] == 0:
                    ans = max(ans, len(words[i]) * len(words[j]))
        return ans

File: 最大单词长度乘积.cpp
class Solution {
public:
    int maxProduct(vector<string>& words) {
        int n = words.size(), ans = 0;
        vector<int> mask(n);
        for (int i = 0; i < n; ++i) {
            for (char c : words[i]) {
                mask[i] |= (1 << (c - 'a'));
            }
        }
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < n; ++j) {
                if ((mask[i] & mask[j]) == 0) {
                    ans = max(ans, (int)words[i].size() * (int)words[j].size());
                }
            }
        }
        return ans;
    }
};
